File: android/src/main/kotlin/io/sourcya/playx_3d_scene/core/shape/sphere/model/Sphere.kt
package io.sourcya.playx_3d_scene.core.shape.sphere.model

import io.sourcya.playx_3d_scene.core.shape.common.material.model.Material
import io.sourcya.playx_3d_scene.core.shape.common.model.Position
import io.sourcya.playx_3d_scene.core.shape.common.model.Shape
import io.sourcya.playx_3d_scene.core.shape.common.model.convertJsonToPosition
import io.sourcya.playx_3d_scene.utils.getMapValue

class Sphere(
    id:Int? = null,
    centerPosition: Position? = null,
    material: Material? = null,
    val radius : Float? = null,
    val  stacks: Int?,
    val slices: Int?,

    ) : Shape(id,centerPosition , null, material) {

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Sphere) return false

        if (radius != other.radius) return false
        if (stacks != other.stacks) return false
        if (slices != other.slices) return false

        return super.equals(other)
    }

    override fun hashCode(): Int {
        var result = super.hashCode()
        result = 31 * result + (radius?.hashCode() ?: 0)
        result = 31 * result + (stacks?.hashCode() ?: 0)
        result = 31 * result + (slices?.hashCode() ?: 0)
        return result
    }

    companion object {
        fun fromMap(map: Map<String?, Any?>?): Sphere? {
            if (map == null) return null

            return Sphere(
                id =getMapValue("id", map),
                centerPosition = convertJsonToPosition(getMapValue("centerPosition", map)),
                material = Material.fromJson(getMapValue("material", map)),
                radius = (getMapValue<Double>("radius", map))?.toFloat(),
                stacks = getMapValue("stacks", map),
                slices = getMapValue("slices", map),
            )
        }
    }
}

File: lib/src/models/shapes/sphere.dart
import 'package:playx_3d_scene/src/models/shapes/shape.dart';

/// An object that represents a cube shape to be rendered.
class Sphere extends Shape {
  /// The radius of the constructed sphere.
  double radius;

  ///The number of stacks for the sphere.
  int? stacks;

  ///The number of slices for the sphere.
  int? slices;

  Sphere(
      {required super.id,
      required super.centerPosition,
      required this.radius,
      this.stacks,
      this.slices,
      super.normal,
      super.material});

  @override
  Map<String, dynamic> toJson() => {
        'id': id,
        'centerPosition': centerPosition?.toJson(),
        'normal': normal?.toJson(),
        'radius': radius,
        'stacks': stacks,
        'slices': slices,
        'material': material?.toJson(),
        'shapeType': 3
      };

  @override
  String toString() {
    return 'Sphere(id: $id, radius: $radius, centerPosition: $centerPosition)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;

    return other is Sphere &&
        other.id == id &&
        other.radius == radius &&
        other.centerPosition == centerPosition;
  }

  @override
  int get hashCode => id.hashCode ^ radius.hashCode ^ centerPosition.hashCode;
}
